spec.html

```diff
@@ -11059,6 +11060,26 @@ <h1>Examples of Automatic Semicolon Insertion</h1>
       <p>In the circumstance that an assignment statement must begin with a left parenthesis, it is a good idea for the programmer to provide an explicit semicolon at the end of the preceding statement rather than to rely on automatic semicolon insertion.</p>
     </emu-clause>
   </emu-clause>
+  <emu-clause id="sec-hazards-of-automatic-semicolon-insertion">
+    <h1>Hazards of Automatic Semicolon Insertion</h1>
+    <em>This section is non-normative.</em>
+    <p>ECMAScript programs can be written in a style with very few semicolons by relying on automatic semicolon insertion. However, as described above, semicolons are not inserted at every newline, and some of these cases which do not have an automatically inserted semicolon can be counter-intuitive or confusing.</p>
+
+    <p>As new syntactic features are added to ECMAScript, additional cases requiring explicit semicolons emerge over time. As such, consistently explicit semicolon use is recommended.</p>
+
+    <p>The term &ldquo;<dfn>automatic semicolon insertion hazard</dfn>&rdquo; refers, informally, to a place where a developer may expect a semicolon to be inserted, but according to the rules described above, it is not. The rest of this section describes a number of automatic semicolon insertion hazards in this version of ECMAScript.</p>
+    <emu-clause id="sec-asi-hazards-in-statement-lists">
+      <h1>Automatic Semicolon Insertion Hazards in Statement Lists</h1>
+      <p>In a |StatementList|, many |StatementListItem|s end in semicolons, which may be omitted using automatic semicolon insertion. As a consequence of the rules above, at the end of a line ending an expression, a semicolon will be required if the following line begins with any of the following:
+      <ul>
+        <li><strong>An opening parenthesis (<code>(</code>)</strong>. Without a semicolon, the two lines together will be treated as a function call.</li>
+        <li><strong>An opening square bracket (<code>[</code>)</strong>. Without a semicolon, the two lines together will be treated as property access, rather than an array literal.</li>
+        <li><strong>A template string (<code>`</code>)</strong>. Without a semicolon, the two lines together will be interpreted as tagged template literal, with the previous expression as the function tag.</li>
+        <li><strong>Unary <code>+</code> or <code>-</code></strong>. Without a semicolon, the two lines together will be interpreted as a usage of the corresponding binary operator.</li>
+        <li><strong>A RegExp literal</strong>. Without a semicolon, the two lines together may be parsed instead as a division, for example if the RegExp has flags.</li>
```

Comment:
I think that one case is lacking here. Should it mention about `[no line terminator here]` syntax?

```js
function f() {
    return // "ASI Hazards" here. It does unexpected ASI.
        100;
}
```

Reply:
This shouldn't be used as it returns undefined.

Reply:
This shouldn't be used as it returns undefined.

Reply:
This is a different category of problem, the issue is not ASI not being invoked, but that `return` has a form that takes 0 operands, similar to why:

```mjs
outer: for (let x of y) {
   while (true) {
      continue
        outer;
   }
}
```

does not continue the `outer` loop.

The existing examples are places where adding code can have unintended affects due to surrounding environment.

```mjs
/hello/i.test(greeting) || sayHi()
```

adding test around it like the following

```mjs
let greeting = 'hello'
/hello/i.test(greeting) || sayHi()
```

would change the meaning of the line with the RegExp literal to become a binary division operation.

Reply:
Doesn't it include `incrementor` and other operators ?

Reply:
@sendilkumarn example?

Reply:
@bmeck This is a well-known ASI trap because people don't write `;` as expected to not insert semicolon automatically. Indeed mechanism is different, but it's an ASI hazard. I feel unfair if "ASI Hazards" section describes about "don't apply ASI unexpectedly" cases but doesn't describe "apply ASI unexpectedly" cases. I felt unfair because I felt that it hides unnecessary information to deprecate semicolon-less style. Thanks.

Reply:
@mysticatea the example is not a hazard in the sense of adding code changes the parsing of a different line. The intent of this PR is not to deprecate any coding style, but to provide a list of problems and recommendation that would avoid problems with additive changes to source text caused by ASI.

Reply:
@bmeck People from C-like languages are taught that these languages are “whitespace-insensitive.” That is, adding line breaks should not alter the meaning of code.

For example,

```js
  return JSON.stringify(processData(extractData(JSON.parse(input))));
```

A developer may try to reformat the code like this:

```js
  return
    JSON.stringify(processData(extractData(JSON.parse(input))));
```

Here, adding code (whitespace) can have unintended affects.

This is why I consider it a hazard.

Reply:
@bmeck The problem with increment opertator is that

`let x=100
let y=101
x
++
y
console.log(x,y)
`

the above will print 100,102

The ambiguity of applying the operator is handled.

Reply:
but when you replace `++` with `<<` or`**` it applies the changes on x rather than y

Reply:
added list of places where `[no LineTerminator here]` is used, explained why this is not considered in same category or a hazard.

Reply:
But isn't this important to show the explicit behavior here between various operand ?

the `updateexpression` is too generic there. It should specifically mention postfix operators (since it behaves differently for shiftwise operator)

Reply:
The only productions of UpdateExpression containing `[no LineTerminator here]` are the ones you mention. I don't think a change is needed since the section is explicitly talking about optional operands and `[no LineTerminator here]`.

Reply:
But when we say it as an `updateexpression`  isn't it also mentions `unaryexpression` (is `lefthandsideexpression` a correct term here ?